### PR TITLE
[BEAM-714] Converts Datastore to use AutoValue

### PR DESCRIPTION
**Pull request by jkff**
This style saves a good amount of code and is easier to mimic in new connectors.

I have a few more changes like this lined up.

R: @dhalperi 
CC: @jbonofre

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java [hunk 1]
```diff
@@ -222,16 +224,23 @@
...
-    @Nullable
-    private final String projectId;
+    @Nullable abstract String getProjectId();
+    @Nullable abstract Query getQuery();
+    @Nullable abstract String getNamespace();
+    abstract int getNumQuerySplits();
 
-    @Nullable
-    private final Query query;
+    public abstract String toString();
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java]: can this possibly be required? That function is on every object.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java [hunk 2]
```diff
@@ -345,25 +354,12 @@ static RunQueryRequest makeRequest(Query query, @Nullable String namespace) {
     }
 
     /**
-     * Note that only {@code namespace} is really {@code @Nullable}. The other parameters may be
-     * {@code null} as a matter of build order, but if they are {@code null} at instantiation time,
-     * an error will be thrown.
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java]: move to L226?

Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java]: Does `.build()` actually verify the non-null parameters?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java (hunk 1 above)
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java]: AutoValue only generates a non-trivial toString() if I specify this.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java (hunk 2 above)
Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java]: build() is an implementation detail, sufficient validation happens in validate().

**Comment by dhalperi** [comment]
Btw can you post a diff of the javadoc output?

**Comment by jkff** [comment]
To have a javadoc, I added the javadoc plugin to pom.xml.

javadoc diff for the Read transform: https://gist.github.com/jkff/9128e4d33b7defba493df9fdbdd5c365/revisions - it looks almost the same, except methods are abstract and the protected constructor now appears.

I did not change the Write transform because there's no gain to be had from AutoValue in it because it currently has just 1 parameter (project id).

**Comment by dhalperi** [comment]
Hmm, the javadoc change _ought to_ be unnecessary, though I'll be darned if I can figure out maven properly. It should be inherited from parents, as we've made Javadoc on-by-default and disabled in certain legacy modules that were not yet ready.

E.g., the old configuration does produce javadoc for GCP IOs: http://beam.incubator.apache.org/learn/sdks/javadoc/0.2.0-incubating/

**Comment by dhalperi** [comment]
Aside from the Javadoc config change, LGTM.

**Comment by jkff** [comment]
Removed javadoc config change - see https://issues.apache.org/jira/browse/BEAM-729

**Comment by vikkyrk** [comment]
This is nice, Thanks!. LGTM

**Comment by jkff** [comment]
Looks like something's wrong with the build itself, mvn clean verify passes locally for me.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java [hunk 3]
```diff
@@ -224,10 +224,12 @@
      */
     static final int QUERY_BATCH_LIMIT = 500;
 
-    @Nullable abstract String getProjectId();
-    @Nullable abstract Query getQuery();
-    @Nullable abstract String getNamespace();
-    abstract int getNumQuerySplits();
+    protected Read() {}
```
Comment (dhalperi) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java]: Purpose of this function? To restrict visibility, plausibly for Javadoc?

Comment (jkff) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/datastore/DatastoreV1.java]: Yeah, though this didn't work and it's still visible in the javadoc. And I can't make it private cause then AutoValue constructors won't work. Removed.

**Comment by jkff** [comment]
PTAL, build passed.

**Pull request closed without merging**